base/set.jl

```diff
@@ -178,6 +178,49 @@ function unique(f, C)
     out
 end
 
+"""
+    unique!(f, A::AbstractVector)
+
+Selects one value from `A` for each unique value produced by `f` applied to
+elements of `A` , then return the modified A.
+
+# Examples
+```jldoctest
+julia> unique!(x -> x^2, [1, -1, 3, -3, 4])
+3-element Array{Int64,1}:
+ 1
+ 3
+ 4
+
+julia> unique!(n -> n%3, [5, 1, 8, 9, 3, 4, 10, 7, 2, 6])
+3-element Array{Int64,1}:
+ 5
+ 1
+ 9
+
+julia> unique!(iseven, [2, 3, 5, 7, 9])
+2-element Array{Int64,1}:
+ 2
+ 3
+```
+"""
+function unique!(f, A::AbstractVector)
+    seen = Set{eltype(A)}()
```

Comment:
That is not going to work our so well.

Consisder:
Current behavour of `unique`


```
julia> unique(x-> x % 2 ==0 ? :even : :odd, [1,2,3,4,2,2,1])
2-element Array{Int64,1}:
 1
 2
```

vs for the this implementation of `unique!`

```
julia> unique!(x-> x % 2 ==0 ? :even : :odd, [1,2,3,4,2,2,1])
ERROR: MethodError: Cannot `convert` an object of type Symbol to an object of type Int64
Closest candidates are:
  convert(::Type{T<:Number}, ::T<:Number) where T<:Number at number.jl:6
  convert(::Type{T<:Number}, ::Number) where T<:Number at number.jl:7
  convert(::Type{T<:Integer}, ::Ptr) where T<:Integer at pointer.jl:23
  ...
Stacktrace:
 [1] setindex!(::Dict{Int64,Nothing}, ::Nothing, ::Symbol) at ./dict.jl:373
 [2] push!(::Set{Int64}, ::Symbol) at ./set.jl:48
 [3] new_unique!(::getfield(Main, Symbol("##3#4")), ::Array{Int64,1}) at ./REPL[8]:9
 [4] top-level scope at none:0
```
```

Reply:
I think this is fine now.
I feel like there is an optimisation that could be done to remove the `Set{Any}()` (like in `map`)
but that it is not required in this PR, since this works the same way as `unique` (and it might not be required at all, benchmarking should be done)

Reply:
So ultimately this is fine by now isn't it?

Reply:
AFAIK it is fine,
but you'll have to wait for someone with merge permissions to do a real review/merge

Reply:
I don't know why travis-ci build test is failing repeatedly.Any advice would be helpful